### ISSUE TITLE
new type of package dependence: dependence on executables only

### ISSUE THREAD

**Issue by notestaff** [conda/conda#3333]
When package  A depends on package B, but only needs from package B some executable files/scripts (and not libraries or include files), then package B can be installed in a separate conda environment.   Then any conflicts between B and its dependencies, and any other packages needed by A, can be completely avoided.  The separate conda environment would be put under the main conda environment (say under etc/env_B ).   For each executable in package B, say etc/env_B/bin/myprog , we create an executable script in the main conda environment (  bin/myprog ) which temporarily activates B's conda environment etc/env_B , then runs the original executable in etc/env_B/bin/myprog, then deactivates B's conda environment.

Thus, for example, if the package A needs python 3, it can still depend on a package that needs python 2, as long as it only needs executables from that package and not libraries.

To implement this, a new type of 'requirements' would be added to meta.xml; say, 'requirements-executables' .

**Comment by kalefranz** [comment]
Wow, just coming across this.  Ends up being very similar to https://github.com/conda/conda/issues/3912

Going to close this, because #3912 is being rolled out over time (was partly in 4.3, more in 4.4, probably won't be 4.5 until it's made more general)

**Comment by notestaff** [comment]
@kalefranz Are you sure it's the same as #3912?  It seems to me quite different.  In #3912, "An application may request that it be installed in a private named environment".  Here, an application specifies the nature of its dependence on other packages: just as 'Host:' dependencies were separated from 'Build:' dependencies, where Host dependencies are things you need to run and Build dependencies are things you need to be present, this suggestion would separate out Run dependencies that just need to be executed, from things that need to be present (libraries and includes).  I often find myself manually implementing this suggestion by installing some tool into a separate environment and then putting in a shim script into my main environment's bin/ directory, which temporarily activates the separate environment and runs the tool.  This let's the tool be installed into an environment unconstrained by anything else, so it gets the latest possible versions of its dependencies and has no conflicts.

**Comment by mingwandroid** [comment]
> Thus, for example, if the package A needs python 3, it can still depend on a package that needs python 2

In this case both the python 3 (env_A) and python 2 (env_B) envs contain a file called `bin/python`, OK, you've put them in separate envs, but if I understand your idea correctly you then suggest overwriting `bin/python` in env_A with a script. That would break `env_A` wouldn't it?

**Comment by notestaff** [comment]
@mingwandroid "you then suggest overwriting bin/python in env_A with a script" no, nothing is ever overwritten.  I want to install pkg_B, which requires python2, into env_A, which has python3.  Currently, I can't.  But, suppose I know that the only way I will use pkg_B is by running executables it places into bin/ .   Then I can install pkg_B and its dependencies, including python2, into a separate env_B, created just for pkg_B; and for each executable that pkg_B puts into env_B/bin, say env_B/bin/prog_B, create a corresponding script env_A/bin/prog_B, which temporarily activates env_B and then calls env_B/bin/prog_B with the same command line and returns its return code.   This way, when I activate env_A, I can use all the executables from pkg_B.

Now, suppose I want to install into my env_A some pkg_A which has a runtime dependency on pkg_B (and therefore on python2).  Currently, I can't.  But if pkg_A only needs to run executables from pkg_B, and not use any libraries or include files, it can specify that information in meta.yaml, by putting its dependence on pkg_B into a separate subsection of the requirements, say run-executables rather than run.  Then, when I ask conda to install pkg_A into env_A, it can install pkg_B into env_B as described above, putting the shim scripts for pkg_B's binaries into env_A/bin; and then proceed to install pkg_A into env_A, where it should work correctly.  But this all would be transparent to the user, who would just see the installation of pkg_A into env_A succeed, where currently it would fail.

Would this cause problems?

**Comment by notestaff** [comment]
@mingwandroid   Another way of doing the same thing would be:
Any package can be automatically converted to a "shim" version which has no dependencies,
and installs only shims into bin/.  So, if you have a package pkg_A,  you can convert pkg_A into pkg_A_executables_only where for each prog_A which pkg_A installs into bin/prog_A, pkg_A_executables_only installs into bin/prog_A a script that, when first run, creates a new environment pkg_A_env (if not yet created), activates it, and runs bin/prog_A from it, with the same command line, and returns its return code.  Any package that depends only on executables from pkg_A can define as runtime dependency pkg_A_executables_only instead of pkg_A .   And a user who only needs executables from pkg_A can install pkg_A_executables into any environment without worrying about conflicts.
But it would be better if, in case of conflict, conda automatically did this.

**Comment by mingwandroid** [comment]
I'm *very* tired and will need to study what you wrote when less so, but let's say for the sake of argument that prog_B *is* python, what then?

**Comment by notestaff** [comment]
@mingwandroid Then no luck.  Can’t put into one environment two different versions of a package, or two different packages that both include bin/prog_X.  But right now, you often can’t put into one environment two different packages that don’t have such conflicts between themselves, but do have conflicts between their dependencies.  With this approach, such packages can effectively coexist in one environment.

**Comment by notestaff** [comment]
@mingwandroid Happy new year! :)  When you have a minute, could you tell me if my explanations above make sense, or if more clarifications would help?

**Comment by notestaff** [comment]
@mingwandroid Thanks for your analysis.  From your comments, it seems I have not explained the scheme well enough, because some of the problems you point out can't in fact happen.
May I try one more time?

"the shared libs would be present in the build-time host env" -- we're talking about run-time requirements only, so why does the build-time host env matter?

"argv[0] is different" -- it's not; I'm sorry if I didn't explain things clearly.  Let me try again.  I want to install into env_A some pkg_B, which conflicts with something already installed in env_A.  But, the _only_ way I'll use pkg_B is by running an executable prog_B which pkg_B installs into bin/.  So, I do 'conda create -n env_B pkg_B'; and then create a "shim" script envs/env_A/bin/prog_B , which reads

#!/bin/bash

set -eu -o pipefail

. /miniconda_root_path/etc/profile.d/conda.sh
conda activate env_B
prog_B "$@"

The last line invokes envs/env_B/bin/prog_B, which sees envs/env_B/bin/prog_B as argv[0].  It does _not_ see envs/env_A/bin/prog_B as argv[0], or as anything else; as far as prog_B is concerned, it is being run in a completely standard way inside env_B.  If prog_B looks for files relative to its executable, it will find them inside env_B in the expected places.


"your scheme cannot work in the case of any program existing in both envs and having the same pathname" -- not so; env_B, which is automatically created by conda, can have any paths that env_A has with no problem.   E.g. env_A can have envs/env_A/bin/python which is python2, and env_B can have envs/env_B/bin/python which is python3.  The _only_ executable added to env_A is the shim script envs/env_A/bin/prog_B .  Only pkg_B provides bin/prog_B, so the shim won't conflict with anything in env_A.   envs/env_B/bin/python (python3) cannot be _run_ directly when env_A is active, but it can _exist_ in env_B regardless of envs/env_A/bin/python also existing.  envs/env_B/bin/python just needs to be available for the shim envs/env_A/bin/prog_B to run after temporarily activating env_B.

"this feature useful only in a very limited set of circumstances" -- it's quite common for a package to work only with Python2 or only with Python3.  This feature lets you install package A (which requires Python2) and package B (which requires Python3) into the same environment; or, to distribute your own package C which depends on both package A and package B; or, to distribute your own package D which depends on package A, and have package D installable into either Python2 or Python3 environments.   All these scenarios currently lead to irresolvable conda conflicts.

Please let me know if more clarifications would help.

**Comment by notestaff** [comment]
@mingwandroid (I've made some edits to the above comment so please read the github web version not the emailed notification version).

**Comment by mingwandroid** [comment]
I believe that I fully understand your idea yes, and also that my concerns stand.

> "the shared libs would be present in the build-time host env" -- we're talking about run-time requirements only, so why does the build-time host env matter?

The ability to build packages always matters. Without packages we have nothing. This was only a minor concern though and please ignore it as it just muddies the waters of this discussion.

>  The only executable added to env_A is the shim script envs/env_A/bin/prog_B

The shim script *needs* to be called `python`, so in this example, it overwrites env_A's `python2` with a shim that points at env_B's `python3`. By calling the programs `prog_A` and `prog_B` in your description you neatly sidestep this messy reality, but still the reality exists.

> "argv[0] is different" it's not; I'm sorry if I didn't explain things clearly

Let's run through the 2 possibilities:

1. No conflict, everything can be installed into env_A. In this case, prog_B's full path is envs/env_A/bin/prog_B.
2. Conflict, conda switches to this behaviour. In this case prog_B's full path is envs/env_B/bin/prog_B.
Now in 99% (fraction pulled from nowhere) I agree that this will not matter, but in the 1% of cases where it does matter it will be problematic. Let's take an example of any MSYS2 'posix' executable. The location of the `msys2` dll is queried from the operating system via `GetModuleFilenameEx` and that location is used as the 'root' location for an entire emulated posix system.
This comes back to similar discussions I've had about 'virtual envs'. If it is easy for a program to detect it's not running in the same execution environment as a 'non-virtual env' then it can vary its behaviour based upon this determination and if it does so then there's a possibility that things will go wrong.

My core point is that this would require an awful lot of changes to `conda` and make the codebase a lot harder to work with. Therefore I would not like the code to be in `conda` itself, so the only way forward I can see is via extension points in `conda` and a custom project that makes use of these.

**Comment by notestaff** [comment]
@mingwandroid "The shim script needs to be called python" -- I've implemented this scheme by hand many times, and never had a shim script called python.  So we must be talking about different things.  Consider two packages from the bioconda channel: dxpy (needs python2) and snakemake (needs python3).  I want to make a new package P that has runtime dependencies on the "dx" command from dxpy and the "snakemake" command from snakemake.  To install P into an env, say env_A, conda would install dxpy and P into env_A normally, install snakemake into a separate snakemake_env, and install a shim envs/env_A/bin/snakemake .  Then the code in P can run "dx" and "snakemake" commands. Where is the need for a shim script called python, or for overwriting anything?

If P _itself_ needed both python2 and python3, there'd be a genuine conflict.  But P doesn't need python at all; it just needs the "dx" and "snakemake" commands.  So there's no real conflict, yet current conda doesn't permit P.  If I want to make P I'll have to rewrite it to not use either "dx" or "snakemake", for no good reason.

"it is easy for a program to detect it's not running in the same execution environment as a 'non-virtual env'" -- how can it?  E.g. above, snakemake_env is just another env, and env_A/bin/snakemake activates snakemake_env before running snakemake_env/bin/snakemake.  How can snakemake_env/bin/snakemake know there even exists another env, besides the current snakemake_env?  And how can the existence of another env affect what snakemake does? 

"Conflict, conda switches to this behaviour" -- only _if_ P's author has explicitly specified that this behavior is safe.  When P is tested by conda-build, this behavior will be tested, and the package won't be made if something fails.  So, with P, I might have to skip the Windows build; but right now I can't build P at all.

"this would require an awful lot of changes to conda" -- maybe not.  Shim version of any package can be [generated on-the-fly](https://github.com/conda/conda/issues/3333#issuecomment-449489074).  Instead of a separate runtime-exec-only section of dependencies, make the convention that a runtime dependency on "snakemake-exonly" denotes a runtime dependency on the shim version of package "snakemake".  Treat "snakemake-exonly" dependency as "snakemake" dependency, except that if conda install hits a conflict, retry with the shim versions of packages, auto-generating them as needed.

**Comment by mingwandroid** [comment]
> "The shim script needs to be called python" -- I've implemented this scheme by hand many times, and never had a shim script called python

You are proposing a general mechanism here based upon only what you have needed (so far). You've not been able to convince me that this isn't a huge flaw in your idea. What if *my* project needs two Python interpreters?

> "it is easy for a program to detect it's not running in the same execution environment as a 'non-virtual env'" -- how can it?

On Linux, via `readlink("/proc/self/exe")`
On macOS, via `NSGetExecutablePath()`
On Windows via `GetModuleFilename{Ex,}()`

I already gave you a real-world example of where a whole ecosystem of programs (and programs we currently provide, the MSYS2 ones) modify their behaviour based upon the location of a DLL, so this is not some theoretical concern.

> "this would require an awful lot of changes to conda" -- maybe not

I believe it would require a *huge* amount of brittle and hard to test changes to `conda` and to `conda-build`. If you feel otherwise then make the PRs to see for yourself. Whether these are generated 'on the fly' or not isn't important, what is is the amount of code and extra special casing logic needed.

I do not think I'm going to convince you though, so I'm not going to continue to try to do that. I stand by my assertion that the only sensible way to implement this is via some extension system. My considered opinion is that conda was not designed for your use-case and hacking it in there would complicate and degrade the quality of the code.

Perhaps @msarahan or @kalefranz can chime in to give a more definitive statement on this subject?

**Comment by notestaff** [comment]
@mingwandroid @msarahan Thanks for your comments, will give it a try.
